File: searchForARange.py
class Solution:
    # @param A, a list of integers
    # @param target, an integer to be searched
    # @return a list of length 2, [index1, index2]
    def searchRange(self, A, target):
        low = 0
        high = len(A) - 1

        # find lower bound, first target
        lowerBound = -1
        while low <= high:
        	mid = (low + high) / 2
        	if A[mid] == target and (mid == low or mid - 1 >= 0 and A[mid-1] < target):
        		lowerBound = mid
        		break
        	elif A[mid] < target:
        		low = mid + 1
        	else:
        		high = mid - 1

        # find higher bound, last target
        higherBound = -1
        low = 0
        high = len(A) - 1
        while low <= high:
        	mid = (low + high) / 2
        	if A[mid] == target and (mid == high or mid + 1 <= len(A) and A[mid+1] > target):
        		higherBound = mid
        		break
        	elif A[mid] <= target:
        		low = mid + 1
        	else:
        		high = mid - 1

        return [lowerBound, higherBound]

A = [5, 7, 7, 8, 8, 10]
target = 10
s = Solution()
print s.searchRange(A, target)

File: JAVA/searchForARange.java
public class Solution {
    public int[] searchRange(int[] A, int target) {
        //Binary Search Part
        int low = 0, high = A.length - 1, pos = -1, mid;
        while (low <= high){
            mid = (low + high) / 2;
            if (A[mid] == target){
                pos = mid;
                break;
            }
            if (A[mid] > target)
                high--;
            else
                low++;
        }
        
        int[] range = new int[2];
        range[0] = pos;
        range[1] = pos;
        
        if (pos == -1)
            return range;
            
        //else extend & get range;
        while(range[0] > 0 && A[range[0]-1] == target)
            range[0] -= 1;
        while(range[1] < A.length-1 && A[range[1]+1] == target)
            range[1] += 1;
        return range;
    }
}

// worst case: O(n)

// Try use binary search find lowerbound and upperbound, with worst case O(n)